### PR TITLE
Issue #26 sign-extend int8/int16/int24/int32 when widening to int64

### PR DESCRIPTION
**Pull request by karmakaze**

### CODE REVIEW

File: mysql/resultset.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ func (p RowData) ParseText(f []*Field) ([]interface{}, error) {
 		if isNull {
 			data[i] = nil
 		} else {
-			isUnsigned = (f[i].Flag&UNSIGNED_FLAG > 0)
+			isUnsigned := f[i].Flag & UNSIGNED_FLAG != 0
```
Comment (karmakaze) [inline mysql/resultset.go]: small style change--advantage being that the var isUnsigned can't be read without having being properly initialized

File: replication/row_event.go [hunk 2]
```diff
@@ -276,7 +276,7 @@ func (e *RowsEvent) Decode(data []byte) error {
 }
 
 func (e *RowsEvent) decodeRows(data []byte, table *TableMapEvent, bitmap []byte) (int, error) {
-	rows := make([]interface{}, e.ColumnCount)
+	row := make([]interface{}, e.ColumnCount)
```
Comment (karmakaze) [inline replication/row_event.go]: I did a double take reading `rows` initialized with ColumnCount elements--renaming seems clearer.

File: replication/row_event.go [hunk 3]
```diff
@@ -342,18 +342,21 @@ func (e *RowsEvent) decodeValue(data []byte, tp byte, meta uint16) (v interface{
...
-		v = int64(binary.LittleEndian.Uint16(data))
+		v = int64(int16(binary.LittleEndian.Uint16(data)))
 	case MYSQL_TYPE_INT24:
 		n = 3
-		v = int64(FixedLengthInt(data[0:3]))
+		u32 := uint32(FixedLengthInt(data[0:3]))
+		if u32 & 0x00800000 != 0 {
+			u32 |= 0xFF000000
+		}
+		v = int64(int32(u32))
```
Comment (karmakaze) [inline replication/row_event.go]: Special case, first 'sign-extend' by setting upper 8 bits if high-bit of 24 is set (i.e. negative).

File: replication/row_event.go [hunk 4]
```diff
@@ -342,18 +342,21 @@ func (e *RowsEvent) decodeValue(data []byte, tp byte, meta uint16) (v interface{
...
 	case MYSQL_TYPE_LONGLONG:
-		//em, maybe overflow for int64......
 		n = 8
 		v = int64(binary.LittleEndian.Uint64(data))
```
Comment (karmakaze) [inline replication/row_event.go]: re-interpreting uint64 as int64 doesn't lose bits. later casting back to uint64 will give original unsigned numeric value, so comment safely removed.

**Comment by siddontang** [comment]
Cool, @karmakaze , I will see it later, Thank you.

**Comment by siddontang** [comment]
Hi @karmakaze , can you add some tests?

**Comment by karmakaze** [comment]
I can see about adding some tests over the weekend. This library has been working out pretty well so far. Performance so far hasn't been an issue, so just looking at correctly handling all the datatypes we're using.

**Comment by karmakaze** [comment]
Added tests, put common implementation in mysql/parse_binary.go used by mysql/resultset.go and replication/row_event.go

**Comment by siddontang** [comment]
LGTM, Thank you @karmakaze

**Pull request merged**